lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -199,6 +199,162 @@ static void printNoIOOp(Operation *op, OpAsmPrinter *printer) {
+                                      OperationState *state) {
+  OpAsmParser::OperandType ptrInfo;
+  SmallVector<OpAsmParser::OperandType, 4> indicesInfo;
+  spirv::StorageClass resultStorageClass;
+  Type type;
+  llvm::SMLoc indicesLocation;
+  int32_t index;
+  Type indicesType = parser->getBuilder().getIndexType();
+
+  if (parser->parseOperand(ptrInfo) ||
+      parser->getCurrentLocation(&indicesLocation) ||
+      parser->parseOperandList(indicesInfo, OpAsmParser::Delimiter::Square) ||
+      parser->parseColonType(type) ||
+      parser->resolveOperand(ptrInfo, type, state->operands) ||
+      parser->resolveOperands(indicesInfo, indicesType, state->operands)) {
+    return failure();
+  }
+
+  if (!indicesInfo.size()) {
+    return parser->emitError(indicesLocation,
+                             "expected at least one index for spv.AccessChain");
+  }
+
+  auto ptrType = type.dyn_cast<spirv::PointerType>();
+  if (!ptrType) {
+    return parser->emitError(
+               indicesLocation,
+               "expected a pointer to composite type, but provided ")
+           << type;
+  }
+
+  resultStorageClass = ptrType.getStorageClass();
+  auto resultType = ptrType.getPointeeType();
+
+  for (uint64_t i = 0; i < indicesInfo.size(); ++i) {
+    if (auto cType = resultType.dyn_cast<spirv::CompositeType>()) {
+      index = 0;
+      if (resultType.isa<spirv::StructType>()) {
+        Operation *op = state->operands[i + 1]->getDefiningOp();
+        if (!op) {
+          return parser->emitError(
+              indicesLocation, "index must be an integer constant to access "
+                               "element of StructType");
+        }
+        auto resultPair = extractValueFromConstOp(op, index);
+        if (!resultPair.second) {
+          return parser->emitError(indicesLocation, resultPair.first);
+        }
+        if (index < 0 ||
+            static_cast<uint64_t>(index) >= cType.getNumElements()) {
+          return parser->emitError(indicesLocation, "index ")
+                 << index << " out of bounds for " << type;
+        }
+      }
+      resultType = cType.getElementType(index);
+    } else {
+      return parser->emitError(indicesLocation,
+                               "cannot extract from non-composite type ")
+             << resultType << " with index " << index;
+    }
+  }
+
+  // Result type must be a pointer with the same StorageClass.
+  state->addTypes(spirv::PointerType::get(resultType, resultStorageClass));
+  return success();
+}
+
+static void print(spirv::AccessChainOp op, OpAsmPrinter *printer) {
+  *printer << spirv::AccessChainOp::getOperationName() << ' ' << *op.base_ptr()
+           << '[';
+  printer->printOperands(op.indices());
+  *printer << "] : " << op.base_ptr()->getType();
+}
+
+static LogicalResult verify(spirv::AccessChainOp accessChainOp) {
+  auto resultType = accessChainOp.base_ptr()
+                        ->getType()
+                        .dyn_cast<spirv::PointerType>()
```

Comment:
We need to check the `dyn_cast`'ed type is actually not null before getting its pointee type. In verifier we cannot assume too much. :)

Reply:
I see, thanks.